mitchspano/sfdx-scan-pull-request
# Report issues with check-runs instead of only comments, and remove comments that are resolved

TS port of the original PR #33

- Be able to switch the mode of reporting from comments (current) and to check runs
    - What is [check runs](https://github.blog/2018-12-14-introducing-check-runs-and-annotations/)? It is the built in mechanism in GitHub pull requests to add annotations from scanning tools (ie, linters). In my opinion it works a bit better than comments as it doesn't produce notifications, they are softer and show as a status on the PR
- Remove resolved comments from the PR
    - Unfortunately, the REST API cannot be used to **resolve** comments, that can only be done via the GraphQL mutations API but it doesn't have cross compatible IDs
    - Added in a prefix to each comment so they can be safely deleted - not the most reliable method
- Fixes bug that made it so the diff scanning never worked (more detail in comment)

Structurally, all of the functionality for generating comments and check-runs have been moved to their own files.

Happy to break up this PR if needed

## Check run preview
![image](https://user-images.githubusercontent.com/314500/226084975-0d2549a9-d2bb-420c-be86-8a5df242daee.png)

To test, you can use the following action:

```yaml
    - name: SFDX Scanner
      uses: mattisdada/sfdx-scan-pull-request@ts-port
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
        report-mode: check-runs # or swap to `comments` to toggle between them
```

## Examples

Comments reporter:
- https://github.com/mattisdada/simple-lambda-router-ts/pull/3

Check runs reporter:
- https://github.com/mattisdada/simple-lambda-router-ts/pull/2

0.1.8 release: (this PR was trying to show the existing bug I mentioned in this PR, but there seems to be multiple issues in 0.1.8, so it's only reported one issue, the action failed)
- https://github.com/mattisdada/simple-lambda-router-ts/pull/4

Video showing the comments being deleted after resolved: (hard to show in a PR because the comment is deleted)
- https://www.youtube.com/watch?v=ds5BEyiexLc
